File: bridge/src/main/kotlin/com/dashmrl/jsbridge/service/Service.kt
package com.dashmrl.jsbridge.service

import android.util.Log
import com.dashmrl.jsbridge.Proxy
import com.dashmrl.jsbridge.callback.JSCallback
import com.dashmrl.jsbridge.util.JSON
import com.dashmrl.jsbridge.model.Request

/**
 * Author       xinliu
 * Date         10/22/17
 * Time         10:31 AM
 * used to handle the call from JS environment
 */
abstract class Service<in T, out R> {
    internal fun handleJSCall(requestStr: String, proxy: Proxy) {
        Log.d("ServiceName", requestStr)
        val request = JSON.parse<Request<T>>(requestStr)
        handle(request.param, JSCallback(proxy, request.uuid))
    }

    abstract fun handle(data: T, callback: JSCallback<R>)

}

File: bridgejs/src/core/service.ts
import { RemoteCallback } from "./callbacks";
import { Request } from "./models";
import { Proxy } from "./proxy";
abstract class Service<T, R> {
    constructor(readonly name: string) { }
    /**
     * handleNativeRequest
     */
    public handleNativeRequest(request: Request<T>, proxy: Proxy) {
        this.handle(request.param, new RemoteCallback(request.uuid, proxy));
    }

    public abstract handle(param: T, callback: RemoteCallback<R>): void;
}
class FallbackService extends Service<any, any> {
    public handle(param: any, callback: RemoteCallback<any>): void {
        callback.falied(`service not found`);
    }
}

export {
    Service,
    FallbackService,
};
